[스터디 상세보기] 스터디상세보기 뷰 기능 구현

no1msh opened the pull request:
## 😋 작업한 내용

- 동일한 뷰(프로필 이미지 + 스터디 성공률)가 재사용 되는 것을 고려하여 커스텀 뷰로 제작
- 스터디 정보(제목, 내용 등) 표시
- 버튼 제작


## 🙏 PR Point

MVVM 패턴을 사용하기에 필요한 ViewModel과 LiveData에 대한 이해도가 부족합니다!!
개인적으로 공부하고 있지만 피드백 주시면 좀 더 수월하게 공부할 것 같습니다. 감사합니다!

구현한 이 페이지의 경우 데이터가 동적으로 바뀌어지지 않는데 이 경우에도 라이브데이터가 꼭 필요한지 궁금합니다.
MVVM 패턴의 뷰모델이 뷰를 참조하지 않게 하려고 하는 이유 때문일까요?


## 👍 관련 이슈

<!-- 작업한 이슈번호를 # 뒤에 붙여주세요. -->

- Resolved : #9

no1msh committed:
Merge remote-tracking branch 'origin/AN/feature/7-home-inflate' into AN/feature/9-studydetail-inflate

no1msh committed:
feat: 티어 컬러 리소스 추가

no1msh committed:
feat: strings 리소스 추가

no1msh committed:
feat: 스터디 참여자 UI 모델 추가

no1msh committed:
feat: 티어 UI 모델 추가

no1msh committed:
feat: 버튼으로 사용될 셀렉터 구현

no1msh committed:
Merge remote-tracking branch 'origin/AN/feature/8-createstudy-inflate' into AN/feature/9-studydetail-inflate

# Conflicts:
#	android/app/src/main/res/values/strings.xml

no1msh committed:
feat: 전체적인 스터디 상세보기 페이지 뷰 구현

no1msh committed:
feat: 프로필 이미지 커스텀 뷰 구현

1. 원형으로 자른 이미지
2. 원형 프로그레스 바

no1msh committed:
feat: String 리소스 추가

no1msh committed:
feat: 스터디 참가자 리스트 어댑터 구현

no1msh committed:
feat: 스터디 참가자 목록 뷰홀더 구현

no1msh committed:
feat: 스터디 참가자 목록 아이템 뷰 구현

no1msh committed:
feat: 스터디 UI 모델 구현

no1msh committed:
feat: 스터디 상세보기 뷰모델 구현

1. 스터디와 스터디 참가자들을 상태값으로 가지고 있습니다.
2. 더미데이터들을 포함시켰습니다.

no1msh committed:
feat: 스터디 상세보기 액티비티 구현

inseonyun commented on android/app/src/main/java/com/created/team201/presentation/studyList/CustomProfileImage.kt:
```diff
@@ -0,0 +1,48 @@
+            .load(image)
+            .circleCrop()
+            .into(profileImage)
+        invalidate()
+        requestLayout()
+    }
+
+    fun setSuccessRate(successRate: Int) {
+        studySuccessRate.progress = successRate
+        invalidate()
+        requestLayout()
+    }
+
+    fun setTier(tierName: String) {
+        val tier = Tier.of(tierName)
+        studySuccessRate.progressTintList = ColorStateList.valueOf(context.getColor(tier.color))
+        invalidate()
+        requestLayout()
```
모든 뷰에서 CustomProfileImage를 set하게 되었을 때 invalidate와 requestLayout을 요청하고 있어요. xml 코드를 확인해보면 위 세 개의 레이아웃은 고정 dp를 갖고 있고, visible한 상태이기 때문에 위와 같은 함수 호출은 굳이?일 거 같다는 생각이 드는데, 각 함수 내에서 위 함수들을 호출 해준 이유가 있나요?
반달이 알고 있는 invalidate와 requestLayout 함수는 어떤 작업을 수행하나요?

no1msh replied:
헉 그대로 옮겨 적다보니 다시 체크하지 않았던 것 같습니다!

`invalidate()`는 뷰를 무효화하는 역할을 합니다. 글자나 색상 등 크기 변화는 없이 단순히 뷰의 속성 등이 변경되어 다시 그려야하는 경우 View 를 다시 그리기 위해 호출하는 메소드입니다.

`requestLayout()`은 만약 뷰의 크기 변화가 발생할 경우 레이아웃의 배치도 달라질 수 있기 때문에 해당 메소드를 호출함으로써 뷰들의 크기 측정부터 다시하게 됩니다.

써니가 말한대로 지금 크기와 뷰의 visible 상태 등.. 변화될 염려가 없기 때문에 해당 함수 두개는 삭제하도록 하겠습니다!

inseonyun commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyDetail.kt:
```diff
@@ -0,0 +1,47 @@
+
+    private fun initStudyParticipantsList() {
+        val studyPeopleAdapter = StudyParticipantsAdapter()
+        binding.rvStudyDetailStudyPeople.setHasFixedSize(true)
```
👍

inseonyun commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyDetailViewModel.kt:
```diff
@@ -0,0 +1,92 @@
+
+class StudyDetailViewModel : ViewModel() {
+    private val _study: MutableLiveData<StudyUIModel> = MutableLiveData()
+    val study: LiveData<StudyUIModel> get() = _study
```
> 구현한 이 페이지의 경우 데이터가 동적으로 바뀌어지지 않는데 이 경우에도 라이브데이터가 꼭 필요한지 궁금합니다.

굳이 변하지 않는 데이터를 LiveData로 생성할 필요는 없다고 생각합니다.
나중에는 굳이 프로퍼티로 둘 필요도 없게 되겠죠?
따라서 서버와 통신 후 받게 되는 데이터를 통해서 보여주게 되고 이는 변하는 값이 아니게 됩니다.
반달은 왜 위와 같이 코드 작성을 하였나요? 꼭 LiveData가 있어야 MVVM은 아닙니다

no1msh replied:
MVVM에 대한 이해도가 낮아 생긴 일 같습니다.
DataBinding, LiveData, ViewModel(AAC) 모두 ViewModel이 View를 참조하지 않겠다는 목적을 가진 MVVM을 도와주는 역할을 하지 필수적인 역할을 하지 않는다는걸 깨달았습니다!
수정하겠습니다!

s9hn replied:
dataBinding은.. 다소 필수일지도..

inseonyun commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyParticipantsAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+    ListAdapter<StudyParticipant, StudyPeopleViewHolder>(diffCallBack) {
+
+    init {
+        setHasStableIds(true)
```
고유한 id를 갖지 않는 어댑터에서 setHasStableIds를 true로 주게 되었을 때에도 성능 개선이 되나요?

no1msh replied:
`setHasStableIds`는 표시할 아이템이 이전에 표시한 아이템과 같을 때 굳이 `onBindViewHolder()`를 호출하지 않도록 아이템마다 고유 ID를 가지고 구분하는 것으로 알고 있습니다.
스터디 참여자들의 경우 애초에 같은 스터디 참여자가 두번 이상 나올 경우가 없기 떄문에 삭제하도록 하겠습니다!

inseonyun commented on android/app/src/main/java/com/created/team201/presentation/studyList/uiModel/Tier.kt:
```diff
@@ -0,0 +1,37 @@
+import androidx.annotation.ColorRes
+import com.created.team201.R
+
+enum class Tier(@ColorRes val color: Int) {
```
어노테이션 사용 👍

inseonyun reviewed (approved):
파이팅

s9hn commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyDetail.kt:
```diff
@@ -0,0 +1,47 @@
+import com.created.team201.presentation.common.BindingActivity
+
+class StudyDetail : BindingActivity<ActivityStudyDetailBinding>(R.layout.activity_study_detail) {
+    private val viewModel: StudyDetailViewModel by viewModels()
```
여러 뷰모델을 공유하게 되면 어떤 문제가 발생할까요?(Malibin style)

no1msh replied:
아항 뷰모델을 하나씩 추가할 때마다 `xxxViewModel` 형식의 이름을 지어주게 될텐데 
지금의 경우엔 `viewModel`이라는 변수명을 사용하고 있어서 추후 변경할 불상사가 있군요!
수정해두겠습니다 군리빈

s9hn commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyDetail.kt:
```diff
@@ -0,0 +1,47 @@
+        val studyPeopleAdapter = StudyParticipantsAdapter()
+        binding.rvStudyDetailStudyPeople.setHasFixedSize(true)
+        binding.rvStudyDetailStudyPeople.adapter = studyPeopleAdapter
+        studyPeopleAdapter.submitList(viewModel.studyParticipants.value)
```
submitList는 실시간으로 값을 업데이트하는 liveData와 찰떡궁합입니다. 
그러나 해당코드는 submitList의 이점을 살릴 수 없어보입니다

no1msh replied:
일단은 정적 데이터를 표시하므로 라이브데이터를 다시 빼주었습니다. 
그치만 군리빈의 말 이해 되었습니다! 후에 다른 페이지 구현할 때 참고해보겠습니다!

s9hn commented on android/app/src/main/res/layout/activity_study_detail.xml:
```diff
@@ -0,0 +1,212 @@
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            app:title="스터디 상세보기"
```
추출바랍니다!

s9hn commented on android/app/src/main/res/layout/activity_study_detail.xml:
```diff
@@ -0,0 +1,212 @@
+            app:layout_constraintTop_toBottomOf="@id/tv_study_detail_study_people_title"
+            app:spanCount="2"
+            tools:itemCount="4"
+            tools:listitem="@layout/item_study_detail_study_people" />
```
orientation의 명시는 디폴트가 버티컬이라 사용하지 않은걸까요?

s9hn commented on android/app/src/main/res/layout/custom_profile_image.xml:
```diff
@@ -0,0 +1,31 @@
+            android:layout_gravity="center"
+            android:indeterminate="false"
+            android:max="100"
+            android:progressBackgroundTint="#626362"
```
컬러리소스 추출은 안되나요?

no1msh replied:
당연 되죠 ㅎㅎ

s9hn commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyDetail.kt:
```diff
@@ -0,0 +1,47 @@
+    private fun initViewModel() {
+        binding.viewModel = viewModel
+        binding.lifecycleOwner = this
+        viewModel.getStudyInformation()
```
해당함수는 두가지일을 하고있는 것 같습니다!

s9hn commented on android/app/src/main/java/com/created/team201/presentation/studyList/StudyParticipantsAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    override fun onBindViewHolder(holder: StudyPeopleViewHolder, position: Int) {
+        holder.bind(getItem(position))
```
getItem(position) vs currentList[position]

no1msh replied:
오홍... 코드 까보니 둘다 `AsyncListDiffer` 인스턴스의 `currentList`를 참조하는 거군요.
근데 둘이 어찌되던 같은 값을 가져오는거 아닌가요?
제가 놓치는 부분이 있다면 알려주세요. 군리빈 ㅠ

s9hn replied:
ㅋㅋ맞음 같은거임 ㅋ.ㅋ

s9hn reviewed (approved):
고생고생티비쓰~

no1msh committed:
refactor: 불필요한 invalidate(), requestLayout() 함수 삭제

no1msh committed:
refactor: 불필요한 liveData 제거

no1msh committed:
refactor: 불필요한 setHasStableIds 제거

no1msh committed:
refactor: viewModel 인스턴스 이름 변경

- viewModel -> studyDetailViewModel

no1msh committed:
refactor: 스터디 상세보기 페이지 앱바 제목 스트링 리소스로 추출

no1msh committed:
refactor: 리사이클러뷰의 orientation을 명시

no1msh committed:
feat: dm 버튼 ContentDescription 추가

no1msh committed:
refactor: 커스텀 프로필 이미지 프로그레스 바 색상 리소스 참조로 변경

no1msh committed:
feat: 스터디 참여하기 버튼을 눌렀을 떄 버튼이 바뀌게끔 변경

no1msh committed:
refactor: CustomProfileImage를 좀더 범용성 있게 변경

1. 기존 도메인에 심하게 국한된 요소 이름을 변경하였습니다.
2. xml 상에서 데이터 바인딩이 아닌 직접 값을 넣어도 적용되게끔 변경하였습니다.

no1msh committed:
Merge remote-tracking branch 'origin/develop' into AN/feature/9-studydetail-inflate

no1msh committed:
Merge branch 'develop' into AN/feature/9-studydetail-inflate

no1msh committed:
Merge remote-tracking branch 'origin/AN/feature/9-studydetail-inflate' into AN/feature/9-studydetail-inflate

no1msh committed:
Merge remote-tracking branch 'origin/develop' into AN/feature/9-studydetail-inflate

# Conflicts:
#	android/app/src/main/res/values/strings.xml

no1msh committed:
chore: 코루틴 의존성 추가

no1msh committed:
Merge remote-tracking branch 'origin/develop' into AN/feature/9-studydetail-inflate

# Conflicts:
#	android/app/build.gradle
#	android/app/src/main/res/values/arrays.xml

no1msh committed:
Merge remote-tracking branch 'origin/develop' into AN/feature/9-studydetail-inflate

# Conflicts:
#	android/app/build.gradle
#	android/app/src/main/res/values/arrays.xml
#	android/app/src/main/res/values/attrs.xml
#	android/app/src/main/res/values/strings.xml

no1msh committed:
chore: 불필요한 String 리소스 제거
